docs/en_US/NAS/Serialization.rst

```diff
@@ -0,0 +1,52 @@
+
+Serialization is implemented as a combination of `json-tricks <https://json-tricks.readthedocs.io/en/latest/>`_ and `cloudpickle <https://github.com/cloudpipe/cloudpickle>`_. Essentially, it is json-tricks, that is a enhanced version of Python JSON, enabling handling of serialization of numpy arrays, date/times, decimal, fraction and etc. The difference lies in the handling of class instances. Json-tricks deals with class instances with ``__dict__`` and ``__class__``, which in most of our cases are not reliable (e.g., datasets, dataloaders). Rather, our serialization deals with class instances with two methods:
+
+1. If the class / factory that creates the object is decorated with ``nni.trace``, we can serialize the type / factory function, along with the parameters, such that the instance can be re-instantiated.
+2. Otherwise, cloudpickle is used to serialize the object into a binary.
+
+The recommendation is, unless you are absolutely certain that there is no problem and extra burden to serialize the object into binary, always add ``nni.trace``. In most cases, it will be more clean and neat, and enables possibilities such as mutation of parameters (will be supported in future).
+
+.. warning::
+
+    **What will happen if I forget to "trace" my objects?**
+
+    It is likely that the program can still run. NNI will try to serialize the untraced object into a binary. If might fail in complicated cases (e.g., circular dependency). Even if it succeeds, the result might be a substantially large object. For example, if you forgot to add ``nni.trace`` on ``MNIST``, the MNIST dataset object wil be serialized into binary, which will be dozens of megabytes because the object has the whole 60k images stored inside. You might see warnings and even errors when running experiments. To avoid such issues, the easiest way is to always remember to add ``nni.trace`` to non-primitive objects.
+
+To trace a function or class, users can use decorator like,
+
+.. code-block:: python
+
+    @nni.trace
+    class MyClass:
+        ...
+
+Inline trace that traces instantly before being used is also acceptable: ``nni.trace(MyClass)(parameters)``.
+
+Assuming a class ``cls`` is already traced, when it is serialized, its class type along with initialization parameters will be dumped. As the parameters are possibly class instances (if not primitive types like ``int`` and ``str``), their serialization will be a similar problem. We recommend decorate them with ``nni.trace`` as well. In other words, ``nni.trace`` should be applied recursively if necessary.
```

Comment:
it would be interesting to automatically add `nni.trace` on necessary objects, possibly using syntax checking technique as plugin :)